### PR TITLE
Add MapStream for easier streaming over maps

### PR DESCRIPTION
**Pull request by cjkent**

### CODE REVIEW

File: modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java [hunk 1]
```diff
@@ -0,0 +1,391 @@
...
+  private final Stream<Map.Entry<K, V>> underlying;
+
+  /**
+   * Returns a stream over the entries in the map.
+   *
+   * @param map a map
+   * @return a stream over the entries in the map
+   */
+  public static <K, V> MapStream<K, V> of(Map<K, V> map) {
+    return new MapStream<>(map.entrySet().stream());
```
Comment (jodastephen) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: Missing `empty()` factory.

File: modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java [hunk 2]
```diff
@@ -0,0 +1,391 @@
...
+  /**
+   * Transforms the entries in the stream by applying a mapper function to each key and value.
+   *
+   * @param mapper  a mapper function whose return values are included in the new stream
+   * @param <R>  the type of elements in the new stream
+   * @return a stream containing the values returned from the mapper function
+   */
+  public <R> Stream<R> map(BiFunction<? super K, ? super V, ? extends R> mapper) {
+    return underlying.map(e -> mapper.apply(e.getKey(), e.getValue()));
+  }
```
Comment (jodastephen) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: Add `mapToValues()` and `mapToKeys()`. These would be useful if you need to filter by key but then only need the values.

File: modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+   * Returns an immutable map built from the entries in the stream.
+   * <p>
+   * The keys must be unique or an exception will be thrown. Duplicate keys can be handled by using
+   * {@link #collect(Collector)} and {@code Collectors.toMap}.
+   *
+   * @return an immutable map built from the entries in the stream
+   */
+  public ImmutableMap<K, V> toImmutableMap() {
+    return underlying.collect(Guavate.toImmutableMap(e -> e.getKey(), e -> e.getValue()));
+  }
```
Comment (jodastephen) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: Could we just have `toMap()`? Would it be clear enough?

File: modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+import java.util.stream.IntStream;
+import java.util.stream.LongStream;
+import java.util.stream.Stream;
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * A stream implementation which adds methods for manipulating keys and values when streaming over map entries.
+ */
+public class MapStream<K, V> implements Stream<Map.Entry<K, V>> {
```
Comment (jodastephen) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: Class should be `final`?

File: modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java [hunk 5]
```diff
@@ -0,0 +1,391 @@
...
+
+  @Override
+  public void close() {
+    underlying.close();
+  }
+
+  //--------------------------------------------------------------------------------------------------
+
+  private static <K, V> Map.Entry<K, V> entry(K k, V v) {
+    return new AbstractMap.SimpleEntry<>(k, v);
```
Comment (jodastephen) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: `SimpleImmutableEntry` not `SimpleEntry`?

File: modules/collect/src/test/java/com/opengamma/strata/collect/MapStreamTest.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    assertThat(result).isEqualTo(expected);
+  }
+
+  public void toMapDuplicateKeys() {
+    assertThrowsIllegalArg(() -> MapStream.of(map).mapKeys(k -> "key").toMap());
+  }
+
+  public void toImmutableMapDuplicateKeys() {
+    assertThrowsIllegalArg(() -> MapStream.of(map).mapKeys(k -> "key").toImmutableMap());
+  }
```
Comment (jodastephen) [inline modules/collect/src/test/java/com/opengamma/strata/collect/MapStreamTest.java]: Test coverage is probably quite low, but other methods are not that interesting. Worth adding a `coverage` test just to deal with them.

**Comment by jodastephen** [comment]
Looks good overall. Will be interested to see how it works in practice.

File: modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java (hunk 2 above)
Comment (cjkent) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: I'm not sure I know what you mean. What signature are you thinking of?

**Comment by cjkent** [comment]
Added `MapStream.forEach` after reviewing usages of `Map.entrySet().stream()` in Strata.

File: modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java (hunk 2 above)
Comment (jodastephen) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: `Stream<V> mapToValues()`

Comment (cjkent) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: What parameters?

Comment (jodastephen) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: None. It is the equivalent of calling `values()` on a `Map`.

Comment (cjkent) [inline modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java]: OK, in which case I don't think the method name should include "map". That's what confused me. Map implies applying a function to the elements. I think `keys()` and `values()` would be better.

But you could achieve the same thing with the existing API like this:

    stream.map((k, v) -> v)...

I'm not sure those 2 methods add enough to justify inclusion at the moment, I've only added the ones I think we really need.

**Pull request merged**